### DISCUSSION TITLE
What is `& full`? (different than xsv full)

### DISCUSSION THREAD

**Discussion by reagle** [saulpw/visidata#1852] [Q&A] [answered]
I understand `& full` was equivalent to `sql:full_outer`, but I've been testing my toy examples with `xsv`:

```
| A  |                                    |
| -- | ---------------------------------- |
| id | name                               |
| 1  | Alice                              |
| 2  | Bob                                |
| 3  | Carol                              |
| \_ | \_<!-- Empty row for alignment --> |

| B                                 |     |
| --------------------------------- | --- |
| id                                | age |
| 1                                 | 25  |
| 2                                 | 30  |
| \_                                | \_  |
| 4                                 | 40  |

| Full Outer Join |       |     |
| --------------- | ----- | --- |
| id              | name  | age |
| 1               | Alice | 25  |
| 2               | Bob   | 30  |
| 3               | Carol | Ø   |
| 4               | Ø     | 40  |
```

but `xsv join --full id eg-full-id-name.tsv id eg-full-id-age.tsv` shows

```
id,name,id,age
1,Alice,1,25
2,Bob,2,30
3,Carol,,
,,4,40
```

**Comment by saulpw** [comment]
I don't understand why xsv would duplicate the `id` column, if `id` is the key?  I don't believe it works this way in SQL.

**Comment by reagle** [comment]
Hrrmmm... there's one other terminal tool I can try, and `--outer` is supposed to be a full outer; it duplicates (but renames) the `id` key column... I forgot to reference the files earlier, but they can be [found here](https://github.com/reagle/visidata2-anki/tree/main/tables).

```
❯ csvjoin --outer -c id eg-full-id-name.tsv eg-full-id-age.tsv

id,name,id2,age
1,Alice,1,25
2,Bob,2,30
3,Carol,,
,,,
,,4,40
```

ChatGPT says this is typical:

<img width="763" alt="SCR-20230414-nsgz" src="https://user-images.githubusercontent.com/596814/232138971-bc64d83f-573d-432b-ad59-b64f7f92e621.png">

**Comment by reagle** [comment]
I don't know why `csvjoin` creates an extra row. 

In any case, perhaps the definition for vd is then: equivalent to SQL:FULL with any redundant key columns removed. (?)

**Reply by saulpw** [reply, marked as answer]
Maybe?  I guess the 'redundant' columns are added because you might need to know which table the row came from.  I'm not sure how rigorous we need to be with VisiData; I think it currently fulfills the spirit of a full join, by combining all the data, so maybe we can just update the documentation.  But maybe we should consider changing the behavior to be more like SQL or xsv or csvjoin.  Opinions welcome.

**Reply by reagle** [reply]
I think I was assuming there was a well-specified model and canonical serializations for SQL joins, which tools would implement/interop on, but perhaps not, or at this level of implementation.
